File: skyvern/browser_extension/protocol.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Literal
from urllib.parse import urlsplit

from skyvern.browser_extension.errors import BrowserExtensionError

LEGACY_PROTOCOL_VERSION = 1
PROTOCOL_VERSION = 2
SUPPORTED_PROTOCOL_VERSIONS = frozenset({LEGACY_PROTOCOL_VERSION, PROTOCOL_VERSION})
EXTENSION_ID = "dhommdmblflboaledbbfkdaapkadphlp"

ALLOWED_OPS = frozenset(
    {
        "debugger.attach",
        "debugger.detach",
        "debugger.send",
        "tabs.create",
        "tabs.remove",
        "tabs.activate",
        "tabs.list",
    }
)

ALLOWED_EVENTS = frozenset(
    {
        "extension.hello",
        "pairing.approved",
        "debugger.event",
        "debugger.detached",
        "scope.tabAdded",
        "scope.tabRemoved",
        "tabs.created",
    }
)

ERROR_CODES = frozenset(
    {
        "AUTH_FAILED",
        "OP_NOT_ALLOWED",
        "TAB_NOT_FOUND",
        "TAB_NOT_SCOPED",
        "RESTRICTED_URL",
        "ATTACH_FAILED",
        "DEBUGGER_DETACHED",
        "CDP_METHOD_NOT_ALLOWED",
        "CDP_ERROR",
        "COMMAND_TIMEOUT",
        "INTERNAL",
    }
)

ALLOWED_CDP_METHOD_PREFIXES = (
    "Accessibility.",
    "Animation.",
    "CSS.",
    "Console.",
    "DOM.",
    "DOMDebugger.",
    "DOMSnapshot.",
    "DOMStorage.",
    "Debugger.",
    "Emulation.",
    "Fetch.",
    "IO.",
    "Input.",
    "Inspector.",
    "Log.",
    "Network.",
    "Overlay.",
    "Page.",
    "Performance.",
    "Profiler.",
    "Runtime.",
    "Security.",
    "Storage.",
    "Target.",
)

DENIED_CDP_METHODS = frozenset(
    {
        "Network.getAllCookies",
        "Network.clearBrowserCookies",
        "Network.clearBrowserCache",
        "Storage.getCookies",
        "Storage.setCookies",
        "Storage.clearCookies",
    }
)

RESTRICTED_URL_PREFIXES = (
    "chrome://",
    "chrome-untrusted://",
    "chrome-extension://",
    "devtools://",
    "edge://",
    "about:",
    "file://",
)

MessageKind = Literal["response", "event", "pong", "ping", "auth.proof", "extension.reset_ack"]


@dataclass(frozen=True, slots=True)
class ParsedMessage:
    kind: MessageKind
    protocol_version: int
    request_id: str | None = None
    ok: bool | None = None
    result: dict | None = None
    error: dict | None = None
    error_code: str | None = None
    error_message: str | None = None
    event: str | None = None
    params: dict | None = None
    client_nonce: str | None = None
    proof: str | None = None
    reset_epoch: str | None = None
    generation: int | None = None
    failed_tab_count: int | None = None


def build_request(request_id: str, op: str, args: dict, *, protocol_version: int = PROTOCOL_VERSION) -> dict:
    if not isinstance(request_id, str) or not request_id:
        raise BrowserExtensionError("Request id must be a non-empty string")
    if not isinstance(op, str) or op not in ALLOWED_OPS:
        raise BrowserExtensionError(f"Operation is not allowed: {op}")
    if not isinstance(args, dict):
        raise BrowserExtensionError("Request args must be an object")
    if protocol_version not in SUPPORTED_PROTOCOL_VERSIONS:
        raise BrowserExtensionError("Unsupported extension protocol version")
    return {"v": protocol_version, "type": "request", "id": request_id, "op": op, "args": args}


def parse_extension_message(raw: str) -> ParsedMessage:
    try:
        payload = json.loads(raw)
    except (json.JSONDecodeError, TypeError) as exc:
        raise BrowserExtensionError("Extension frame is not valid JSON") from exc

    if not isinstance(payload, dict):
        raise BrowserExtensionError("Extension frame must be an object")
    protocol_version = payload.get("v")
    if type(protocol_version) is not int or protocol_version not in SUPPORTED_PROTOCOL_VERSIONS:
        raise BrowserExtensionError("Unsupported extension protocol version")

    frame_type = payload.get("type")
    if frame_type == "response":
        return _parse_response(payload, protocol_version)
    if frame_type == "event":
        return _parse_event(payload, protocol_version)
    if frame_type == "ping":
        return ParsedMessage(kind="ping", protocol_version=protocol_version)
    if frame_type == "pong":
        return ParsedMessage(kind="pong", protocol_version=protocol_version)
    if frame_type == "auth.proof":
        return ParsedMessage(
            kind="auth.proof",
            protocol_version=protocol_version,
            client_nonce=_required_string(payload, "clientNonce"),
            proof=_required_string(payload, "proof"),
        )
    if frame_type == "extension.reset_ack" and protocol_version == PROTOCOL_VERSION:
        reset_epoch = _required_string(payload, "epoch")
        generation = payload.get("generation")
        if type(generation) is not int or generation < 0:
            raise BrowserExtensionError("Extension reset generation must be a non-negative integer")
        ok = payload.get("ok")
        if type(ok) is not bool:
            raise BrowserExtensionError("Extension reset ok must be a boolean")
        failed_tab_count = payload.get("failedTabCount", 0)
        if type(failed_tab_count) is not int or failed_tab_count < 0 or (not ok and failed_tab_count == 0):
            raise BrowserExtensionError("Extension reset failedTabCount must describe failed detachments")
        if ok and failed_tab_count != 0:
            raise BrowserExtensionError("Successful extension reset cannot contain failed detachments")
        return ParsedMessage(
            kind="extension.reset_ack",
            protocol_version=protocol_version,
            reset_epoch=reset_epoch,
            generation=generation,
            ok=ok,
            failed_tab_count=failed_tab_count,
        )
    raise BrowserExtensionError("Unknown extension frame type")


def is_cdp_method_allowed(method: str, params: dict | None = None) -> bool:
    if method in DENIED_CDP_METHODS:
        return False
    if method == "Network.getCookies" and isinstance(params, dict) and "urls" in params:
        return False
    return any(method.startswith(prefix) and len(method) > len(prefix) for prefix in ALLOWED_CDP_METHOD_PREFIXES)


def is_restricted_url(url: str) -> bool:
    normalized_url = url.strip().lower()
    if normalized_url == "about:blank":
        return False
    if normalized_url.startswith(RESTRICTED_URL_PREFIXES):
        return True
    try:
        hostname = urlsplit(normalized_url).hostname
        return hostname is not None and hostname.removesuffix(".") == "chromewebstore.google.com"
    except ValueError:
        return False


def _parse_response(payload: dict, protocol_version: int) -> ParsedMessage:
    request_id = _required_string(payload, "id")
    ok = payload.get("ok")
    if type(ok) is not bool:
        raise BrowserExtensionError("Response ok must be a boolean")

    if ok:
        result = _required_dict(payload, "result")
        return ParsedMessage(
            kind="response", protocol_version=protocol_version, request_id=request_id, ok=True, result=result
        )

    error = _required_dict(payload, "error")
    error_code = _required_string(error, "code")
    error_message = _required_string(error, "message")
    if error_code not in ERROR_CODES:
        raise BrowserExtensionError("Response contains an unknown error code")
    return ParsedMessage(
        kind="response",
        protocol_version=protocol_version,
        request_id=request_id,
        ok=False,
        error=error,
        error_code=error_code,
        error_message=error_message,
    )


def _parse_event(payload: dict, protocol_version: int) -> ParsedMessage:
    event = _required_string(payload, "event")
    if event not in ALLOWED_EVENTS:
        raise BrowserExtensionError("Unknown extension event")
    params = _required_dict(payload, "params")
    if event == "extension.hello":
        reported_version = params.get("protocolVersion")
        if protocol_version == PROTOCOL_VERSION and reported_version != protocol_version:
            raise BrowserExtensionError("Extension hello protocolVersion must match the frame version")
        if reported_version is not None and reported_version != protocol_version:
            raise BrowserExtensionError("Extension hello protocolVersion must match the frame version")
    return ParsedMessage(kind="event", protocol_version=protocol_version, event=event, params=params)


def _required_string(payload: dict, field: str) -> str:
    value = payload.get(field)
    if not isinstance(value, str) or not value:
        raise BrowserExtensionError(f"Extension frame field {field} must be a non-empty string")
    return value


def _required_dict(payload: dict, field: str) -> dict:
    value = payload.get(field)
    if not isinstance(value, dict):
        raise BrowserExtensionError(f"Extension frame field {field} must be an object")
    return value

File: skyvern/browser_extension/extension/protocol.js
export const PAGE_CHANGED_WHILE_RUNNING_MESSAGE =
  "The page changed while the extension operation was running.";
export const PAGE_CHANGED_BEFORE_START_MESSAGE =
  "The page changed before the extension operation started.";

export const PAGE_CHANGE_TIER_R_METHODS = new Set([
  "Page.getFrameTree",
  "Page.getLayoutMetrics",
  "Page.captureScreenshot",
  "Network.getResponseBody",
  "Page.enable",
  "Runtime.enable",
  "Log.enable",
  "Network.enable",
  "Page.setLifecycleEventsEnabled",
  "Page.createIsolatedWorld",
  "Target.setAutoAttach",
  "Page.addScriptToEvaluateOnNewDocument",
]);
export const PAGE_CHANGE_TIER_Q_METHODS = new Set([
  "Runtime.runIfWaitingForDebugger",
  "Emulation.setFocusEmulationEnabled",
  "Emulation.setEmulatedMedia",
]);

export function isPageChangeBootstrap(method, params) {
  if (
    method !== "Runtime.evaluate" ||
    params === null ||
    typeof params !== "object" ||
    Object.getPrototypeOf(params) !== Object.prototype ||
    Object.keys(params).length !== 2 ||
    !Object.hasOwn(params, "expression") ||
    !Object.hasOwn(params, "contextId") ||
    !Number.isInteger(params.contextId) ||
    params.contextId < 0 ||
    typeof params.expression !== "string"
  ) {
    return false;
  }
  const expression = params.expression.trim();
  if (!/^\(\(\) => \{\s+const module = \{\};/.test(expression)) {
    return false;
  }
  if (
    /return new \(module\.exports\.UtilityScript\(\)\)\(globalThis, (?:true|false)\);\s+\}\)\(\);$/.test(
      expression,
    )
  ) {
    return true;
  }
  const injected = expression.match(
    /return new \(module\.exports\.InjectedScript\(\)\)\(globalThis, (\{[^\r\n]*\})\);\s+\}\)\(\);$/,
  );
  if (injected === null) return false;
  try {
    JSON.parse(injected[1]);
    return true;
  } catch {
    return false;
  }
}

export function isPageChangeExempt(method, params) {
  if (method === "Page.addScriptToEvaluateOnNewDocument") {
    return params?.source === "" && params.runImmediately !== true;
  }
  return (
    PAGE_CHANGE_TIER_R_METHODS.has(method) ||
    PAGE_CHANGE_TIER_Q_METHODS.has(method) ||
    isPageChangeBootstrap(method, params)
  );
}

export const LEGACY_PROTOCOL_VERSION = 1;
export const PROTOCOL_VERSION = 2;
export const DEFAULT_BRIDGE_PORT = 19777;
export const BRIDGE_ALARM_NAME = "skyvern-bridge-reconnect";

export const MESSAGE_TYPES = Object.freeze({
  AUTH_CHALLENGE: "auth.challenge",
  AUTH_PROOF: "auth.proof",
  AUTH_OK: "auth.ok",
  REQUEST: "request",
  RESPONSE: "response",
  EVENT: "event",
  PING: "ping",
  PONG: "pong",
  EXTENSION_RESET: "extension.reset",
  EXTENSION_RESET_ACK: "extension.reset_ack",
});

export const OPS = Object.freeze({
  DEBUGGER_ATTACH: "debugger.attach",
  DEBUGGER_DETACH: "debugger.detach",
  DEBUGGER_SEND: "debugger.send",
  DOM_EVALUATE: "dom.evaluate",
  DOM_FILL: "dom.fill",
  SHARE_TAB: "shareTab",
  UNSHARE_TAB: "unshareTab",
  TABS_CREATE: "tabs.create",
  TABS_REMOVE: "tabs.remove",
  TABS_ACTIVATE: "tabs.activate",
  TABS_LIST: "tabs.list",
});

export const EVENTS = Object.freeze({
  EXTENSION_HELLO: "extension.hello",
  PAIRING_APPROVED: "pairing.approved",
  PAIRING_APPROVED_ACK: "pairing.approved_ack",
  DEBUGGER_EVENT: "debugger.event",
  DEBUGGER_DETACHED: "debugger.detached",
  SCOPE_TAB_ADDED: "scope.tabAdded",
  SCOPE_TAB_REMOVED: "scope.tabRemoved",
  TABS_CREATED: "tabs.created",
});

export const ERROR_CODES = Object.freeze({
  AUTH_FAILED: "AUTH_FAILED",
  OP_NOT_ALLOWED: "OP_NOT_ALLOWED",
  TAB_NOT_FOUND: "TAB_NOT_FOUND",
  TAB_NOT_SCOPED: "TAB_NOT_SCOPED",
  RESTRICTED_URL: "RESTRICTED_URL",
  ATTACH_FAILED: "ATTACH_FAILED",
  DEBUGGER_DETACHED: "DEBUGGER_DETACHED",
  CDP_METHOD_NOT_ALLOWED: "CDP_METHOD_NOT_ALLOWED",
  CDP_ERROR: "CDP_ERROR",
  COMMAND_TIMEOUT: "COMMAND_TIMEOUT",
  INTERNAL: "INTERNAL",
});

export const ALLOWED_CDP_PREFIXES = new Set([
  "Accessibility",
  "Animation",
  "CSS",
  "Console",
  "DOM",
  "DOMDebugger",
  "DOMSnapshot",
  "DOMStorage",
  "Debugger",
  "Emulation",
  "Fetch",
  "IO",
  "Input",
  "Inspector",
  "Log",
  "Network",
  "Overlay",
  "Page",
  "Performance",
  "Profiler",
  "Runtime",
  "Security",
  "Storage",
  "Target",
]);

export const DENIED_CDP_METHODS = new Set([
  "Network.getAllCookies",
  "Network.clearBrowserCookies",
  "Network.clearBrowserCache",
  "Storage.getCookies",
  "Storage.setCookies",
  "Storage.clearCookies",
]);

export class ProtocolError extends Error {
  constructor(code, message) {
    super(message);
    this.name = "ProtocolError";
    this.code = code;
  }
}

export function isRestrictedUrl(url) {
  if (typeof url !== "string" || url.length === 0) {
    return false;
  }

  const normalized = url.trim().toLowerCase();
  if (normalized === "about:blank") {
    return false;
  }
  if (
    normalized.startsWith("chrome://") ||
    normalized.startsWith("chrome-untrusted://") ||
    normalized.startsWith("chrome-extension://") ||
    normalized.startsWith("devtools://") ||
    normalized.startsWith("edge://") ||
    normalized.startsWith("about:") ||
    normalized.startsWith("file://")
  ) {
    return true;
  }

  try {
    const hostname = new URL(url).hostname.toLowerCase().replace(/\.$/, "");
    return hostname === "chromewebstore.google.com";
  } catch {
    return false;
  }
}

export function requireArgs(args) {
  if (args === null || typeof args !== "object" || Array.isArray(args)) {
    throw new ProtocolError(
      ERROR_CODES.INTERNAL,
      "Request arguments must be an object.",
    );
  }
  return args;
}

export function requireTabId(value) {
  if (!Number.isInteger(value) || value < 0) {
    throw new ProtocolError(
      ERROR_CODES.TAB_NOT_FOUND,
      "The requested tab was not found.",
    );
  }
  return value;
}

export function protocolErrorEnvelope(error) {
  if (
    error instanceof ProtocolError &&
    Object.values(ERROR_CODES).includes(error.code)
  ) {
    return { code: error.code, message: error.message };
  }
  return {
    code: ERROR_CODES.INTERNAL,
    message: "The extension operation failed.",
  };
}
